# Roll A — Data Loading

In [1]:
import os
from dotenv import load_dotenv

load_dotenv()

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_KEY = os.getenv("SUPABASE_KEY")

print("URL exists:", SUPABASE_URL is not None)
print("KEY exists:", SUPABASE_KEY is not None)

URL exists: True
KEY exists: True


In [ ]:
import pandas as pd
from supabase import create_client

In [ ]:
supabase = create_client(SUPABASE_URL, SUPABASE_KEY)

print("Supabase connected")

Supabase connected


In [4]:

# Kqigi andmete laadimine sales tabelist
all_sales = []
offset = 0
while True:
    response = supabase.table('sales').select('*').range(offset, offset + 999).execute()
    if not response.data:
        break
    all_sales.extend(response.data)
    offset += 1000

df_sales = pd.DataFrame(all_sales)
print("Sales shape:", df_sales.shape)
df_sales.head()


Sales shape: (10118, 12)


,id,sale_id,invoice_id,sale_date,customer_id,product_id,quantity,unit_price,total_price,channel,store_location,payment_method
0,1,1,INV-202301-00001,2023-01-10T00:00:00,2588.0,1274,2,234.79,469.58,pood,Tallinn,kaart
1,2,2,INV-202301-00002,2023-01-16T00:00:00,4338.0,1207,2,241.13,482.26,pood,Pärnu,järelmaks
2,3,3,INV-202301-00003,2023-01-05T00:00:00,4673.0,1264,1,258.46,221.19,pood,Pärnu,järelmaks
3,4,4,INV-202301-00004,2023-01-02T00:00:00,4677.0,1341,3,45.21,135.63,pood,Tartu,sularaha
4,5,5,INV-202301-00005,2023-01-13T00:00:00,2390.0,1284,1,99.57,99.57,pood,Tartu,kaart


In [5]:
# Kqigi andmete laadimine customers tabelist
all_customers = []
offset = 0
while True:
    response = supabase.table('customers').select('*').range(offset, offset + 999).execute()
    if not response.data:
        break
    all_customers.extend(response.data)
    offset += 1000

df_customers = pd.DataFrame(all_customers)
print("Customers shape:", df_customers.shape)
df_customers.head()

Customers shape: (3150, 9)


,customer_id,first_name,last_name,email,phone,city,registration_date,loyalty_tier,birth_year
0,2001,Eha,Aas,eha.aas@telia.ee,+372 8713 1455,Tallinn,2024-02-27,NaN,1973
1,2002,Aivar,Kõiv,aivar.koiv@outlook.com,+372 8943 8684,Haapsalu,2025-01-09,bronze,1988
2,2003,Maris,Rebane,maris.rebane@telia.ee,+372 5918 5726,Tartu,2021-02-03,NaN,1999
3,2004,Jaak,Talvik,jaak.talvik@mail.ee,+372 8554 4232,Tallinn,2023-11-12,silver,1974
4,2005,Raivo,Koppel,raivo.koppel@yahoo.com,+372 5298 4365,Tallinn,2023-05-22,bronze,2004


In [ ]:
df_merge = pd.merge(
    df_sales,
    df_customers,
    on="customer_id",
    how="left"
)

print("Merged shape:", df_merge.shape)

df_merge.head()

Merged shape: (8950, 20)


,id,sale_id,invoice_id,sale_date,customer_id,product_id,quantity,unit_price,total_price,channel,store_location,payment_method,first_name,last_name,email,phone,city,registration_date,loyalty_tier,birth_year
0,1,1,INV-202301-00001,2023-01-10T00:00:00,2588.0,1274,2,234.79,469.58,pood,Tallinn,kaart,Hille,Paju,NaN,+372 5429 0294,Tallinn,2022-07-28,bronze,1997.0
1,2,2,INV-202301-00002,2023-01-16T00:00:00,4338.0,1207,2,241.13,482.26,pood,Pärnu,järelmaks,Merle,Luik,merle.luik@mail.ee,+372 5150 1812,Tallinn,2020-09-22,NaN,1996.0
2,3,3,INV-202301-00003,2023-01-05T00:00:00,4673.0,1264,1,258.46,221.19,pood,Pärnu,järelmaks,Liina,Saar,liina.saar@gmail.com,+372 8809 7990,Tallinn,2020-03-31,silver,1973.0
3,4,4,INV-202301-00004,2023-01-02T00:00:00,4677.0,1341,3,45.21,135.63,pood,Tartu,sularaha,Aili,Pihl,aili.pihl@yahoo.com,+372 8375 4888,Narva,2021-10-08,gold,1972.0
4,5,5,INV-202301-00005,2023-01-13T00:00:00,2390.0,1284,1,99.57,99.57,pood,Tartu,kaart,Triin,Lill,triin.lill@telia.ee,+372 5378 0596,Tartu,2021-04-09,NaN,1996.0


In [52]:
print("Merged DataFrame shape:")
print(df_merge.shape)

print("\nData types:")
print(df_merge.dtypes)

print("\nFirst 5 rows:")
display(df_merge.head())

Merged DataFrame shape:
(8950, 20)

Data types:
id                            int64
sale_id                       int64
invoice_id                      str
sale_date            datetime64[us]
customer_id                 float64
product_id                    int64
quantity                      int64
unit_price                  float64
total_price                 float64
channel                         str
store_location                  str
payment_method                  str
first_name                      str
last_name                       str
email                           str
phone                           str
city                            str
registration_date               str
loyalty_tier                    str
birth_year                  float64
dtype: object

First 5 rows:


,id,sale_id,invoice_id,sale_date,customer_id,product_id,quantity,unit_price,total_price,channel,store_location,payment_method,first_name,last_name,email,phone,city,registration_date,loyalty_tier,birth_year
0,1,1,INV-202301-00001,2023-01-10,2588.0,1274,2,234.79,469.58,pood,Tallinn,kaart,Hille,Paju,NaN,+372 5429 0294,Tallinn,2022-07-28,bronze,1997.0
1,2,2,INV-202301-00002,2023-01-16,4338.0,1207,2,241.13,482.26,pood,Pärnu,järelmaks,Merle,Luik,merle.luik@mail.ee,+372 5150 1812,Tallinn,2020-09-22,NaN,1996.0
2,3,3,INV-202301-00003,2023-01-05,4673.0,1264,1,258.46,221.19,pood,Pärnu,järelmaks,Liina,Saar,liina.saar@gmail.com,+372 8809 7990,Tallinn,2020-03-31,silver,1973.0
3,4,4,INV-202301-00004,2023-01-02,4677.0,1341,3,45.21,135.63,pood,Tartu,sularaha,Aili,Pihl,aili.pihl@yahoo.com,+372 8375 4888,Narva,2021-10-08,gold,1972.0
4,5,5,INV-202301-00005,2023-01-13,2390.0,1284,1,99.57,99.57,pood,Tartu,kaart,Triin,Lill,triin.lill@telia.ee,+372 5378 0596,Tartu,2021-04-09,NaN,1996.0


In [26]:
required_columns = ["customer_id", "sale_date", "total_price", "email"]

for column in required_columns:
    print(column, "exists:", column in df_merge.columns)

customer_id exists: True
sale_date exists: True
total_price exists: True
email exists: True


# ROLL B - Data Cleaning

In [29]:
print("Initial shape:", df_merge.shape)

Initial shape: (10118, 20)


In [31]:
print("Duplicates:", df_merge.duplicated().sum())
df_merge= df_merge.drop_duplicates()
print("After removing duplicates:", df.shape)

Duplicates: 0
After removing duplicates: (8950, 20)


In [11]:
print("NULL values:\n", df.isnull().sum())

NULL values:
 id                      0
sale_id                 0
invoice_id              0
sale_date               0
customer_id           988
product_id              0
quantity                0
unit_price              0
total_price             0
channel                 0
store_location       3462
payment_method          0
first_name            988
last_name             988
email                1944
phone                 988
city                  988
registration_date     988
loyalty_tier         4660
birth_year            988
dtype: int64


In [32]:
df_merge = df_merge.dropna(subset=['customer_id', 'sale_date', 'total_price'])
print("After removing critical NULLs:", df.shape)

After removing critical NULLs: (8950, 20)


In [33]:
df_merge['sale_date'] = pd.to_datetime(df_merge['sale_date'])
print("sale_date type:", df_merge['sale_date'].dtype)

sale_date type: datetime64[us]


In [36]:
print("Negative total_price values:", (df_merge['total_price'] <= 0).sum())
df_merge = df_merge[df_merge['total_price'] > 0]
print("After removing outliers:", df_merge.shape)

Negative total_price values: 0
After removing outliers: (8950, 20)


In [37]:
print("--- CLEANING REPORT ---")
print(f"Final shape: {df_merge.shape}")
print(f"Unique customers: {df_merge['customer_id'].nunique()}")
print(f"Date range: {df_merge['sale_date'].min().date()} to {df_merge['sale_date'].max().date()}")

--- CLEANING REPORT ---
Final shape: (8950, 20)
Unique customers: 2540
Date range: 2023-01-01 to 2026-06-28


# ROLL C - FM kliendisegmenteerimine

In [18]:
today = pd.to_datetime('2026-05-12')

In [40]:
recency = df_merge.groupby('customer_id')['sale_date'].max().reset_index()
recency.columns = ['customer_id', 'last_purchase']
recency['recency_days'] = (today - recency['last_purchase']).dt.days
print(recency.head())


   customer_id last_purchase  recency_days
0       2001.0    2024-11-29           529
1       2004.0    2024-12-19           509
2       2005.0    2024-10-03           586
3       2006.0    2023-11-09           915
4       2007.0    2025-01-30           467


In [41]:
frequency = df_merge.groupby('customer_id')['sale_id'].count().reset_index()
frequency.columns = ['customer_id', 'frequency']
print(frequency.head())

   customer_id  frequency
0       2001.0          2
1       2004.0          2
2       2005.0          4
3       2006.0          1
4       2007.0          1


In [42]:
monetary = df_merge.groupby('customer_id')['total_price'].sum().reset_index()
monetary.columns = ['customer_id', 'monetary']
print(monetary.head())

   customer_id  monetary
0       2001.0    203.92
1       2004.0   1198.56
2       2005.0    959.60
3       2006.0    327.06
4       2007.0    318.63


In [43]:
# Liidan R, F, M kokku
rfm = pd.merge(recency, frequency, on='customer_id')
rfm = pd.merge(rfm, monetary, on='customer_id')

# Skoorid 1-5
rfm['R_score'] = pd.qcut(rfm['recency_days'], q=5, labels=[5,4,3,2,1])
rfm['F_score'] = pd.qcut(rfm['frequency'].rank(method='first'), q=5, labels=[1,2,3,4,5])
rfm['M_score'] = pd.qcut(rfm['monetary'].rank(method='first'), q=5, labels=[1,2,3,4,5])

print(rfm.head())

   customer_id last_purchase  recency_days  frequency  monetary R_score  \
0       2001.0    2024-11-29           529          2    203.92       4   
1       2004.0    2024-12-19           509          2   1198.56       4   
2       2005.0    2024-10-03           586          4    959.60       3   
3       2006.0    2023-11-09           915          1    327.06       1   
4       2007.0    2025-01-30           467          1    318.63       5   

  F_score M_score  
0       1       1  
1       1       4  
2       4       3  
3       1       1  
4       1       1  


In [44]:
# RFM koondskoor
rfm['RFM_score'] = rfm['R_score'].astype(int) + rfm['F_score'].astype(int) + rfm['M_score'].astype(int)

# Segmendid
def segment(score):
    if score >= 13:
        return 'VIP Champions'
    elif score >= 10:
        return 'Loyal'
    elif score >= 7:
        return 'Potential'
    elif score >= 4:
        return 'At Risk'
    else:
        return 'Lost'

rfm['segment'] = rfm['RFM_score'].apply(segment)

# Kokkuvõte
print(rfm['segment'].value_counts())

segment
Potential        759
Loyal            679
At Risk          531
VIP Champions    455
Lost             116
Name: count, dtype: int64


In [45]:
# Kas kõik kliendid said segmendi
print(f"Kokku kliente: {rfm.shape[0]}")
print(f"Segmendita kliente: {rfm['segment'].isnull().sum()}")

# Segmentide jaotus
print("\nSegmentide jaotus:")
print(rfm['segment'].value_counts())

# Protsentides
print("\nProtsentides:")
print(rfm['segment'].value_counts(normalize=True).mul(100).round(1).astype(str) + '%')

Kokku kliente: 2540
Segmendita kliente: 0

Segmentide jaotus:
segment
Potential        759
Loyal            679
At Risk          531
VIP Champions    455
Lost             116
Name: count, dtype: int64

Protsentides:
segment
Potential        29.9%
Loyal            26.7%
At Risk          20.9%
VIP Champions    17.9%
Lost              4.6%
Name: proportion, dtype: str


Kokkuvõte kokku kliente: 2540
Segmendita kliente: 0

VIP Champions 17.9% –  ei ole liiga suur
Potential 29.9% – suur grupp, keda saab kasvatada lojaalseks
Loyal 26.7% – tugev püsiklientide baas
At Risk 20.9% – päris suur grupp, vajab tähelepanu
Lost 4.6% – väike, 
Järeldus:
Potential + Loyal = 56.6% klientidest on perspektiivsed
At Risk kliendid vajaksid  kampaaniat
VIP kliente tasub premeerida et hoida neid

# Roll D - Visualization (Visualiseerimine ja leiud)

In [47]:
import plotly.express as px

In [48]:
# Segmentide klientide arv
segment_data = rfm['segment'].value_counts().reset_index()
segment_data.columns = ['segment', 'klientide_arv']

fig = px.bar(
    segment_data,
    x='segment',
    y='klientide_arv',
    title='Diagramm 1. UrbanStyle segmentide klientide arv',
    labels={'segment': 'Kliendisegment', 'klientide_arv': 'Klientide arv (tk)'},
    color_discrete_sequence=['#009B8D'],
    text_auto=True
)

fig.update_layout(
    plot_bgcolor='#F5F5F5'
)

fig.update_traces(
    textfont_color='#1A1A2E'
)
fig.update_layout(
    height=650,
    width=1200
)
fig.show()

In [ ]:
# Kliendisegmendi ostukäitumine: Päevi viimasest ostust vs kogukulutus
fig = px.scatter(
    rfm,
    x='recency_days',
    y='monetary',
    color='segment',
    size='frequency',
    title='Diagramm 2. UrbanStyle kliendisegmendi ostukäitumine (RFM)',
    labels={
        'recency_days': 'Päevi viimasest ostust',
        'monetary': 'Kogukulutus (EUR)',
        'segment': 'Segment',
        'frequency': 'Ostude arv'
    },
    color_discrete_map={
        'VIP Champions': '#009B8D',
        'Potential': '#AAAAAA',
        'Loyal': '#636EFA',
        'At Risk': '#EF553B',
        'Lost': '#1A1A2E'
    },
    size_max=20
)

fig.update_layout(
    plot_bgcolor='#F5F5F5'
)
fig.update_layout(
    height=650,
    width=1200
)

fig.show()

In [50]:
# Top 10 VIP klienti kogukulutuse järgi
vip_top10 = rfm[rfm['segment'] == 'VIP Champions'].nlargest(10, 'monetary').copy()
vip_top10['customer_id'] = 'Klient ' + vip_top10['customer_id'].astype(str)

fig = px.bar(
    vip_top10,
    x='customer_id',
    y='monetary',
    title='Diagramm 3.Top 10 VIP klienti kogukulutuse järgi',
    labels={'customer_id': 'Klient', 'monetary': 'Kogukulutus (EUR)'},
    color_discrete_sequence=['#009B8D'],
    text_auto=True
)

fig.update_layout(
    plot_bgcolor='#F5F5F5'
)

fig.update_traces(
    textfont_color='#1A1A2E'
)

fig.update_layout(
    height=600,
    width=1200
)

fig.show()

In [54]:
# Kokku kulutused 
total_clients = rfm.shape[0]
total_monetary = rfm['monetary'].sum()


In [ ]:

# VIP analüüs
vip = rfm[rfm['segment'] == 'VIP Champions']
vip_count = vip.shape[0]
vip_monetary = vip['monetary'].sum()
vip_share = (vip_monetary / total_monetary * 100).round(1)

# At Risk analüüs
at_risk = rfm[rfm['segment'] == 'At Risk']
at_risk_count = at_risk.shape[0]
at_risk_monetary = at_risk['monetary'].sum()
at_risk_share = (at_risk_monetary / total_monetary * 100).round(1)

# Keskmine ostusagedus ja kulutus segmentide kaupa
segment_summary = rfm.groupby('segment').agg(
    klientide_arv=('customer_id', 'count'),
    kesk_kulutus=('monetary', 'mean'),
    kesk_sagedus=('frequency', 'mean'),
    kesk_recency=('recency_days', 'mean')
).round(1)

print(f"{'='*40}")
print(f"KLIENDIRAPORT")
print(f"{'='*40}")
print(f"Kokku kliente:         {total_clients}")
print(f"Kokku käive:           {total_monetary:,.0f} EUR")
print(f"{'='*40}")
print(f"VIP kliendid:          {vip_count} ({round(vip_count/total_clients*100, 1)}%)")
print(f"VIP käibeosakaal:      {vip_share}%")
print(f"{'='*40}")
print(f"At Risk kliendid:      {at_risk_count} ({round(at_risk_count/total_clients*100, 1)}%)")
print(f"At Risk käibeosakaal:  {at_risk_share}%")
print(f"{'='*40}")
print(f"\nSegmentide kokkuvõte:")
print(segment_summary)

KLIENDIRAPORT
Kokku kliente:         2540
Kokku käive:           2,676,851 EUR
VIP kliendid:          455 (17.9%)
VIP käibeosakaal:      42.8%
At Risk kliendid:      531 (20.9%)
At Risk käibeosakaal:  7.2%

Segmentide kokkuvõte:
               klientide_arv  kesk_kulutus  kesk_sagedus  kesk_recency
segment                                                               
At Risk                  531         363.2           1.6         748.5
Lost                     116         168.5           1.0         954.1
Loyal                    679        1172.8           3.8         583.3
Potential                759         687.5           2.5         645.5
VIP Champions            455        2519.3           7.7         486.7
